email/pardot.go

```diff
@@ -0,0 +1,189 @@
+}
+
+// PardotClient handles authentication and sending contacts to Pardot.
+type PardotClient struct {
+	businessUnit string
+	clientId     string
+	clientSecret string
+	prospectsURL string
+	tokenURL     string
+	token        *oAuthToken
+	clk          clock.Clock
+}
+
+// NewPardotClient creates a new PardotClient.
+func NewPardotClient(clk clock.Clock, businessUnit, clientId, clientSecret, oauthbaseURL, pardotBaseURL string) (*PardotClient, error) {
+	prospectsURL, err := url.JoinPath(pardotBaseURL, prospectsPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to join upsert path: %w", err)
+	}
+	tokenURL, err := url.JoinPath(oauthbaseURL, tokenPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to join token path: %w", err)
+	}
+
+	return &PardotClient{
+		businessUnit: businessUnit,
+		clientId:     clientId,
+		clientSecret: clientSecret,
+		prospectsURL: prospectsURL,
+		tokenURL:     tokenURL,
+
+		token: &oAuthToken{},
+		clk:   clk,
+	}, nil
+}
+
+// updateToken updates the OAuth token if necessary.
+func (pc *PardotClient) updateToken() error {
+	pc.token.Lock()
+	defer pc.token.Unlock()
+
+	now := pc.clk.Now()
+	if now.Before(pc.token.expiresAt.Add(-5*time.Minute)) && pc.token.accessToken != "" {
+		return nil
+	}
+
+	resp, err := http.PostForm(pc.tokenURL, url.Values{
+		"grant_type":    {"client_credentials"},
+		"client_id":     {pc.clientId},
+		"client_secret": {pc.clientSecret},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to retrieve token: %w", err)
+	}
+	defer resp.Body.Close()
+
+	var respJSON struct {
+		AccessToken string `json:"access_token"`
+		ExpiresIn   int    `json:"expires_in"`
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		err = json.NewDecoder(resp.Body).Decode(&respJSON)
+		if err != nil {
+			return fmt.Errorf("failed to decode token response: %w", err)
+		}
+		pc.token.accessToken = respJSON.AccessToken
+		pc.token.expiresAt = pc.clk.Now().Add(time.Duration(respJSON.ExpiresIn) * time.Second)
+		return nil
+	}
+
+	body, readErr := io.ReadAll(resp.Body)
+	if readErr != nil {
+		return fmt.Errorf("token request failed with status %d; while reading body: %w", resp.StatusCode, readErr)
+	}
+	return fmt.Errorf("token request failed with status %d: %s", resp.StatusCode, body)
+}
+
+// redactEmail replaces all occurrences of an email address in a response body
+// with "[REDACTED]".
+func redactEmail(body []byte, email string) string {
+	return string(bytes.ReplaceAll(body, []byte(email), []byte("[REDACTED]")))
+}
+
+// UpsertEmail sends an email to the Pardot Prospects endpoint, retrying up
+// to 3 times with exponential backoff.
+func (pc *PardotClient) UpsertEmail(email string) error {
+	var finalErr error
+
+	for attempt := 1; attempt <= maxAttempts; attempt++ {
+		err := pc.updateToken()
+		if err != nil {
+			finalErr = err
+		} else {
+			break
+		}
+
+		if attempt < maxAttempts {
+			time.Sleep(core.RetryBackoff(attempt, retryBackoffMin, retryBackoffMax, retryBackoffBase))
+		}
+	}
```

Comment:
All other uses of core.RetryBackoff do so at the very beginning of the loop. This is because RetryBackoff always returns 0 when attempt is 0, so no sleep happens before the first loop. It also allows the sleep to be unconditional, and for the loop to use a simpler `range`.
```suggestion
	for attempt := range maxAttempts {
		time.Sleep(core.RetryBackoff(attempt, retryBackoffMin, retryBackoffMax, retryBackoffBase))
		err = pc.updateToken()
		if err != nil {
			continue
		}
		break
	}
```